itests/org.openhab.core.io.net.tests/src/main/java/org/openhab/core/io/net/tests/ClientFactoryTest.java

```diff
@@ -0,0 +1,189 @@
+/**
+ * Copyright (c) 2010-2023 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.core.io.net.tests;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+import java.net.InetSocketAddress;
+import java.util.Objects;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.jetty.client.HttpClient;
+import org.eclipse.jetty.client.api.ContentResponse;
+import org.eclipse.jetty.http.HttpFields;
+import org.eclipse.jetty.http.HttpMethod;
+import org.eclipse.jetty.http.HttpStatus;
+import org.eclipse.jetty.http.HttpURI;
+import org.eclipse.jetty.http.HttpVersion;
+import org.eclipse.jetty.http.MetaData;
+import org.eclipse.jetty.http2.api.Session;
+import org.eclipse.jetty.http2.api.Stream;
+import org.eclipse.jetty.http2.client.HTTP2Client;
+import org.eclipse.jetty.http2.frames.HeadersFrame;
+import org.eclipse.jetty.util.Promise.Completable;
+import org.eclipse.jetty.websocket.client.WebSocketClient;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import org.openhab.core.io.net.http.HttpClientFactory;
+import org.openhab.core.io.net.http.WebSocketFactory;
+import org.openhab.core.io.net.tests.internal.TestServer;
+import org.openhab.core.io.net.tests.internal.TestStreamAdapter;
+import org.openhab.core.io.net.tests.internal.TestWebSocket;
+import org.openhab.core.test.TestPortUtil;
+import org.openhab.core.test.java.JavaOSGiTest;
+
+/**
+ * Tests for HttpClientFactory and WebSocketFactory implementations.
+ *
+ * @author Andrew Fiddian-Green - Initial contribution
+ */
+@NonNullByDefault
+public class ClientFactoryTest extends JavaOSGiTest {
+    public static final String RESPONSE = "response";
+
+    private static final String CONSUMER = "consumer";
+    private static final String HOST = "127.0.0.1";
+    private static final int WAIT_SECONDS = 10;
+    private static final long CONNECTION_TIMEOUT_MS = TimeUnit.SECONDS.toMillis(WAIT_SECONDS);
+
+    private static @Nullable TestServer server;
```

Comment:
In other tests we use `@NonNullByDefault({})` for fields that are initialized in `@BeforeEach` and `@BeforeAll`. You also don't need the `getServer` method then. I there is nothing else that speaks against using that, I would suggest to do the same here for consistency. Otherwise LGTM.

Reply:
> @NonNullByDefault({})

I tried that but it does not work with static fields.

Reply:
Yup. And I tried it before you. .. :)